Support for reinferred lambdas in GetSymbolInfo and GetDeclaredSymbol

333fred opened the pull request:
We now support exposing reinferred lambda symbols via GetSymbolInfo. As part of this, we have to recreate any local or parameter symbols who's `ContainingSymbol` would have been the original lambda symbol. As part of that, I've created a simple reinferred variable tracker as part of the NullableWalker, that is mutable inside a single pass of the walker and has a hierarchy of previous passes. SemanticModel will now use this for getting reinferred local symbols instead of recreating the symbols itself. @dotnet/roslyn-compiler for review.

Closes https://github.com/dotnet/roslyn/issues/38555 by linking to the original bug, and skips a new test.

333fred commented:
/cc @jasonmalinowski @ryzngard

333fred commented on src/Compilers/CSharp/Portable/Compilation/InitializerSemanticModel.cs:
```diff
@@ -260,7 +260,9 @@ internal override bool TryGetSpeculativeSemanticModelForMethodBodyCore(SyntaxTre
         protected override BoundNode RewriteNullableBoundNodesWithSnapshots(BoundNode boundRoot, Binder binder, DiagnosticBag diagnostics, bool createSnapshots, out NullableWalker.SnapshotManager snapshotManager)
         {
-            return NullableWalker.AnalyzeAndRewrite(Compilation, MemberSymbol, boundRoot, binder, diagnostics, createSnapshots, out snapshotManager);
+            var symbol = boundRoot is BoundPropertyEqualsValue { Property: SourcePropertySymbol { BackingField: var backingField } } ? backingField : MemberSymbol;
```
This is necessary because the initializer's containing symbol is the backing field, not the property.

jcouv commented:
It looks like some legitimate test failures remain.

cston commented on src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs:
```diff
@@ -594,5 +594,23 @@ internal static Location FirstOrNone(this ImmutableArray<Location> items)
+
+        internal static bool SequenceEqual<TElement, TArg>(ImmutableArray<TElement> array1, ImmutableArray<TElement> array2, TArg arg, Func<TElement, TElement, TArg, bool> predicate)
+        {
+            if (array1.IsDefault || array2.IsDefault || array1.Length != array2.Length)
```
>array1.IsDefault || array2.IsDefault [](start = 16, length = 36)

Is it important to allow these cases? Aren't those cases invalid arguments? And if they are not invalid, shouldn't `array1.IsDefault && array2.IsDefault` return `true`?

cston commented on src/Compilers/CSharp/Portable/Symbols/Source/LambdaSymbol.cs:
```diff
@@ -375,13 +375,13 @@ public sealed override bool Equals(Symbol symbol, TypeCompareKind compareKind)
-                && System.Linq.ImmutableArrayExtensions.SequenceEqual(lambda.ParameterTypesWithAnnotations, this.ParameterTypesWithAnnotations, (p1, p2) => p1.Type.Equals(p2.Type))
-                && Equals(lambda.ContainingSymbol, this.ContainingSymbol);
+                && ImmutableArrayExtensions.SequenceEqual(ParameterTypesWithAnnotations, lambda.ParameterTypesWithAnnotations, compareKind,
+                                                          (p1, p2, compareKind) => p1.Equals(p2, compareKind))
```
Can we use `System.Linq.ImmutableArrayExtensions.SequenceEqual(..., ..., IEqualityComparer<> comparer)` instead, with `SymbolEqualityComparer`?

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.ReinferredVariableTracker.cs:
```diff
@@ -0,0 +1,67 @@
+        {
+            internal readonly struct Immutable
+            {
+                private readonly ReinferredVariableTracker? _tracker;
```
>ReinferredVariableTracker? _tracker [](start = 33, length = 35)

Is this immutable? What prevents another call to `SetIReinferredSymbol`?

333fred replied:
I could call the struct `Readonly` instead? The intent is much like `ReadOnlySpan`

---
In reply to: [327278382](https://github.com/dotnet/roslyn/pull/38769#discussion_r327278382) [](ancestors = 327278382)

cston commented on src/Compilers/CSharp/Test/Symbol/Symbols/Source/NullablePublicAPITests.cs:
```diff
@@ -2965,5 +2965,405 @@ void verifyAnnotations(BinaryExpressionSyntax syntax, PublicNullableAnnotation a
+            var o1Ref = innerLambda.DescendantNodes()
+                .OfType<AssignmentExpressionSyntax>()
+                .Skip(1)
+                .First()
```
`ElementAt(1)`

cston commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs:
```diff
@@ -401,22 +401,24 @@ internal void SetTypeWithAnnotations(TypeWithAnnotations newType)
         /// Gets a new local symbol with the given TypeWithAnnotations as the new type. This
         /// type should be identical to the original except for nullability.
```
>nullability [](start = 64, length = 11)

"... and tuple names"?

333fred replied:
It's not intended to be so, we just don't handle tuple names correctly in the walker currently.

---
In reply to: [327301886](https://github.com/dotnet/roslyn/pull/38769#discussion_r327301886) [](ancestors = 327301886)

cston commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs:
```diff
@@ -401,22 +401,24 @@ internal void SetTypeWithAnnotations(TypeWithAnnotations newType)
-        internal LocalSymbol WithAnalyzedType(TypeWithAnnotations analyzedType)
+        internal LocalSymbol WithContainingAndAnalyzedTypes(Symbol containingSymbol, TypeWithAnnotations analyzedType)
         {
+            Debug.Assert(SymbolEqualityComparer.IgnoreNullableAndTupleNames.Equals(ContainingSymbol, containingSymbol));
```
>IgnoreNullableAndTupleNames [](start = 48, length = 27)

Is it a bug that tuple names may be distinct?

333fred replied:
Yes, [https://github.com/dotnet/roslyn/issues/35010](https://github.com/dotnet/roslyn/issues/35010)

---
In reply to: [327302307](https://github.com/dotnet/roslyn/pull/38769#discussion_r327302307) [](ancestors = 327302307)

AlekseyTs commented on src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs:
```diff
@@ -594,5 +594,23 @@ internal static Location FirstOrNone(this ImmutableArray<Location> items)
             return items.IsEmpty ? Location.None : items[0];
         }
+
+        internal static bool SequenceEqual<TElement, TArg>(ImmutableArray<TElement> array1, ImmutableArray<TElement> array2, TArg arg, Func<TElement, TElement, TArg, bool> predicate)
```
>ImmutableArray<TElement> array1 [](start = 59, length = 31)

Should this have a ```this``` modifier?

333fred replied:
For things like SequenceEqual, I prefer the style of having both things be arguments rather than an extension method. If you'd prefer to have it as an extension, reopen and I can change it.

---
In reply to: [327330140](https://github.com/dotnet/roslyn/pull/38769#discussion_r327330140) [](ancestors = 327330140)

AlekseyTs replied:
> For things like SequenceEqual, I prefer the style of having both things be arguments rather than an extension method.

It would be extremely hard to find this helper. It won't show in the completion list after the dot. One would have to be sure it exists and remember this type's name

---
In reply to: [327339134](https://github.com/dotnet/roslyn/pull/38769#discussion_r327339134) [](ancestors = 327339134,327330140)

333fred replied:
Alright, I'll change it.

---
In reply to: [327340766](https://github.com/dotnet/roslyn/pull/38769#discussion_r327340766) [](ancestors = 327340766,327339134,327330140)

AlekseyTs commented on src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs:
```diff
@@ -594,5 +594,23 @@ internal static Location FirstOrNone(this ImmutableArray<Location> items)
             return items.IsEmpty ? Location.None : items[0];
         }
+
+        internal static bool SequenceEqual<TElement, TArg>(ImmutableArray<TElement> array1, ImmutableArray<TElement> array2, TArg arg, Func<TElement, TElement, TArg, bool> predicate)
```
>TArg arg [](start = 125, length = 8)

Minor: It looks like other similar methods in this file put this parameter after the delegate that uses it.

333fred replied:
This ordering was intentional, and I'd argue that the rest of the functions in this file should change. By putting the argument before the lambda, we give the compiler enough info to infer the type of `TArg` before the lambda is typed, meaning that you get actual intellisense in the body.

---
In reply to: [327331056](https://github.com/dotnet/roslyn/pull/38769#discussion_r327331056) [](ancestors = 327331056)

AlekseyTs commented on src/Compilers/Core/Portable/Symbols/SymbolEqualityComparer.cs:
```diff
@@ -24,6 +24,8 @@ public sealed class SymbolEqualityComparer : IEqualityComparer<ISymbol>
         // Internal only comparisons:
         internal readonly static SymbolEqualityComparer ConsiderEverything = new SymbolEqualityComparer(TypeCompareKind.ConsiderEverything);
 
+        internal readonly static SymbolEqualityComparer IgnoreNullableAndTupleNames = new SymbolEqualityComparer(TypeCompareKind.AllIgnoreOptions | TypeCompareKind.IgnoreTupleNames);
```
>IgnoreNullableAndTupleNames [](start = 56, length = 27)

It looks like we already have EqualsIgnoringTupleNamesAndNullability comparer at [https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Portable/Symbols/TypeSymbol.cs#L121](https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Portable/Symbols/TypeSymbol.cs#L121).

AlekseyTs replied:
I think that we should avoid using ```IEqualityComparer<ISymbol>``` within compilers for internal symbol classes. This causes an unnecessary overhead in the form of type checks before we call back into the compiler. Second, we are very likely to stop implementing ISymbol on internal symbols soon.

---
In reply to: [327332447](https://github.com/dotnet/roslyn/pull/38769#discussion_r327332447) [](ancestors = 327332447)

333fred replied:
I'll do this for the set of comparisons introduced in this PR, but I think it's out of scope to change all usages.

---
In reply to: [327335261](https://github.com/dotnet/roslyn/pull/38769#discussion_r327335261) [](ancestors = 327335261,327332447)

AlekseyTs replied:
> I think it's out of scope to change all usages.

Agree. Let's just no make the problem bigger.

---
In reply to: [327350175](https://github.com/dotnet/roslyn/pull/38769#discussion_r327350175) [](ancestors = 327350175,327335261,327332447)

AlekseyTs commented on src/Compilers/Core/Portable/Symbols/SymbolEqualityComparer.cs:
```diff
@@ -24,6 +24,8 @@ public sealed class SymbolEqualityComparer : IEqualityComparer<ISymbol>
         // Internal only comparisons:
         internal readonly static SymbolEqualityComparer ConsiderEverything = new SymbolEqualityComparer(TypeCompareKind.ConsiderEverything);
 
+        internal readonly static SymbolEqualityComparer IgnoreNullableAndTupleNames = new SymbolEqualityComparer(TypeCompareKind.AllIgnoreOptions | TypeCompareKind.IgnoreTupleNames);
```
>TypeCompareKind.AllIgnoreOptions | TypeCompareKind.IgnoreTupleNames [](start = 113, length = 67)

It looks like the set of flags doesn't match the name of the field. Also, ```TypeCompareKind.AllIgnoreOptions``` already ignores everything, so I think adding ```TypeCompareKind.IgnoreTupleNames``` doesn't change anything.

333fred replied:
It was intended to be AllNullable, must have mistyped. I'll remove this declaration entirely.

---
In reply to: [327333301](https://github.com/dotnet/roslyn/pull/38769#discussion_r327333301) [](ancestors = 327333301)

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs:
```diff
@@ -569,6 +556,33 @@ private BoundLambda ReallyBind(NamedTypeSymbol delegateType)
+            diagnostics = DiagnosticBag.GetInstance();
+            cacheKey = ReturnInferenceCacheKey.Create(delegateType, IsAsync);
+
+            // When binding for real (not for return inference), we cannot reuse a body of a lambda
```
>// When binding for real (not for return inference), we cannot reuse a body of a lambda [](start = 12, length = 87)

It feels like this comment belongs in its original place in ```ReallyBind``` rather than in this method which doesn't do anything with body.

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/InitializerSemanticModel.cs:
```diff
@@ -260,7 +260,9 @@ internal override bool TryGetSpeculativeSemanticModelForMethodBodyCore(SyntaxTre
         protected override BoundNode RewriteNullableBoundNodesWithSnapshots(BoundNode boundRoot, Binder binder, DiagnosticBag diagnostics, bool createSnapshots, out NullableWalker.SnapshotManager snapshotManager)
         {
-            return NullableWalker.AnalyzeAndRewrite(Compilation, MemberSymbol, boundRoot, binder, diagnostics, createSnapshots, out snapshotManager);
+            var symbol = boundRoot is BoundPropertyEqualsValue { Property: SourcePropertySymbol { BackingField: var backingField } } ? backingField : MemberSymbol;
```
>var symbol = boundRoot is BoundPropertyEqualsValue { Property: SourcePropertySymbol { BackingField: var backingField } } ? backingField : MemberSymbol; [](start = 12, length = 151)

When ```MemberSymbol``` is a property we want to get its backing field instead? Is this what we are doing here?

333fred replied:
Correct, but this will be removed from the next iteration. We no longer need it as we're no longer assuming that the symbol `NullableWalker` is provided is the containing symbol.

---
In reply to: [327344300](https://github.com/dotnet/roslyn/pull/38769#discussion_r327344300) [](ancestors = 327344300)

AlekseyTs replied:
> Correct, but this will be removed from the next iteration.

Sounds good. Otherwise dependency on bound node feels unnecessary to me, we could jut figure this out from the ```MemberSymbol``` itself.

---
In reply to: [327347067](https://github.com/dotnet/roslyn/pull/38769#discussion_r327347067) [](ancestors = 327347067,327344300)

AlekseyTs commented:
>         internal override LocalSymbol GetAdjustedLocalSymbol(LocalSymbol local, int position)

It feels strange to me that we are dependent on position in this API. Why would declared/inferred type of a local depend on any position or change based on position? What am I missing?

---
Refers to: src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:666 in 708099d. [](commit_id = 708099da280d909e093be09cca26646f5b8412f4, deletion_comment = False)

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:
```diff
@@ -701,6 +681,21 @@ internal override LocalSymbol GetAdjustedLocalSymbol(LocalSymbol local, int posi
 
         private LocalFunctionSymbol GetDeclaredLocalFunction(LocalFunctionStatementSyntax declarationSyntax, SyntaxToken declaredIdentifier)
```
>declaredIdentifier [](start = 121, length = 18)

Can we get this from ```declarationSyntax```?

333fred replied:
Yes, but I was avoiding changing existing code as much as I could here.

---
In reply to: [327350060](https://github.com/dotnet/roslyn/pull/38769#discussion_r327350060) [](ancestors = 327350060)

AlekseyTs replied:
> Yes, but I was avoiding changing existing code as much as I could here.

It looks like there is only one consumer and it passes ```declarationSyntax.Identifier```. Consider getting that below where we actually need it.

---
In reply to: [327351863](https://github.com/dotnet/roslyn/pull/38769#discussion_r327351863) [](ancestors = 327351863,327350060)

333fred commented:
>         internal override LocalSymbol GetAdjustedLocalSymbol(LocalSymbol local, int position)

The only thing that we are dependent on for position here is to know what scope we are in, in order to look up the correct global state efficiently. Currently, the algorithm is: find the local state for a position, which is a binary search. Grab the global state for that position, and then get the reinferred variables. If we were to just directly scan global states, we'd end up searching the same scope potentially multiple times. We could avoid this by being clever with a visited list in the snapshotmanager, but that feels like a lot of work to avoid passing a `position`.

---
In reply to: [534303012](https://github.com/dotnet/roslyn/pull/38769#issuecomment-534303012) [](ancestors = 534303012)

---
Refers to: src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:666 in 708099d. [](commit_id = 708099da280d909e093be09cca26646f5b8412f4, deletion_comment = False)

AlekseyTs commented:
>         internal override LocalSymbol GetAdjustedLocalSymbol(LocalSymbol local, int position)

As we discussed offline, it looks like we should be able to figure out the right position in the snapshot manager from the symbol itself, instead of requiring consumer to pass the right position.

---
In reply to: [534304497](https://github.com/dotnet/roslyn/pull/38769#issuecomment-534304497) [](ancestors = 534304497,534303012)

---
Refers to: src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:666 in 708099d. [](commit_id = 708099da280d909e093be09cca26646f5b8412f4, deletion_comment = False)

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:
```diff
@@ -701,6 +681,21 @@ internal override LocalSymbol GetAdjustedLocalSymbol(LocalSymbol local, int posi
+                    return localFunctionNode.Symbol;
+                }
+
+                return null;
```
>return null; [](start = 16, length = 12)

Should we just fall through to the default handling? It looks like this is what would happen for local symbol.

AlekseyTs commented:
>         internal override LocalSymbol GetAdjustedLocalSymbol(LocalSymbol local, int position)

It looks like this parameter is no longer used. Remove?

---
Refers to: src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:666 in 19ede7f. [](commit_id = 19ede7f9083fa519776ad1d0fb146bc74b3e02df, deletion_comment = False)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs:
```diff
@@ -401,22 +401,24 @@ internal void SetTypeWithAnnotations(TypeWithAnnotations newType)
         /// type should be identical to the original except for nullability.
         /// </summary>
-        internal LocalSymbol WithAnalyzedType(TypeWithAnnotations analyzedType)
+        internal LocalSymbol WithContainingAndAnalyzedTypes(Symbol containingSymbol, TypeWithAnnotations analyzedType)
```
>WithContainingAndAnalyzedTypes [](start = 29, length = 30)

WithContaining**Symbol**AndAnalyzed**Type**?

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs:
```diff
@@ -397,11 +397,12 @@ protected override void VisitSwitchSection(BoundSwitchSection node, bool isLastS
                                 }
-                                else
+
+                                if (local.IsVar)
```
>if (local.IsVar) [](start = 32, length = 16)

Would it make sense to move all code that uses ```variableAccess``` local into this ```if```? It looks like executing it will not have any useful side-effects unless we get into this ```if```.

333fred replied:
This code is no longer changed from existing, will leave as is.

---
In reply to: [327847841](https://github.com/dotnet/roslyn/pull/38769#discussion_r327847841) [](ancestors = 327847841)

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.ReinferredVariableTracker.cs:
```diff
@@ -0,0 +1,67 @@
+                }
+            }
+
+            private readonly Dictionary<Symbol, Symbol> _reinferredVariables;
```
>_reinferredVariables [](start = 56, length = 20)

Are they all local variables? If not, perhaps use "Symbol" term instead. The same goes for the type names.

333fred replied:
There are parameters in here, I'll update the name.

---
In reply to: [327850790](https://github.com/dotnet/roslyn/pull/38769#discussion_r327850790) [](ancestors = 327850790)

AlekseyTs replied:
> There are parameters in here

What about local functions and lambdas?

---
In reply to: [327851162](https://github.com/dotnet/roslyn/pull/38769#discussion_r327851162) [](ancestors = 327851162,327850790)

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.ReinferredVariableTracker.cs:
```diff
@@ -0,0 +1,67 @@
+
+            private ReinferredVariableTracker(ReinferredVariableTracker.Readonly parent)
+            {
+                _reinferredVariables = new Dictionary<Symbol, Symbol>();
```
>new Dictionary<Symbol, Symbol>(); [](start = 39, length = 33)

What comparer would be appropriate for this dictionary? I assume ReferenceEqualityComparer is what we want.

333fred replied:
Reference is what we want, so I didn't set one. I can add a comment.

---
In reply to: [327852656](https://github.com/dotnet/roslyn/pull/38769#discussion_r327852656) [](ancestors = 327852656)

AlekseyTs replied:
> Reference is what we want, so I didn't set one. I can add a comment.

Shouldn't we use ```ReferenceEqualityComparer``` then? Default comparer is not reference comparer.

---
In reply to: [327852930](https://github.com/dotnet/roslyn/pull/38769#discussion_r327852930) [](ancestors = 327852930,327852656)

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -668,12 +670,29 @@ private void TakeIncrementalSnapshot(BoundNode node)
+            if (_reinferredVariablesOpt is object)
+            {
+                updatedLocal = ((SourceLocalSymbol)originalLocal).WithContainingAndAnalyzedTypes(GetUpdatedContainingSymbol(originalLocal), updatedType);
+                if ((object)originalLocal == updatedLocal) return;
```
>return [](start = 59, length = 6)

Do we need to call `SetUpdatedSymbol()` always, even if the value is the original, in case the map was updated previously with a distinct value?

333fred replied:
No longer relevant.

---
In reply to: [327870685](https://github.com/dotnet/roslyn/pull/38769#discussion_r327870685) [](ancestors = 327870685)

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -668,12 +670,29 @@ private void TakeIncrementalSnapshot(BoundNode node)
+        private Symbol GetUpdatedContainingSymbol(Symbol childSymbol) =>
+            _reinferredVariablesOpt?.GetReinferredSymbol(childSymbol.ContainingSymbol) ?? childSymbol.ContainingSymbol;
+
+        private void SetUpdatedSymbol<T>(BoundNode node, T originalSymbol, T updatedSymbol) where T : Symbol
```
><T> [](start = 37, length = 3)

Is the type parameter necessary?

333fred replied:
The type parameter is there to ensure that consumers pass the same original symbol and updated type at compile time. I messed this up once during development so I added harder guarantees.

---
In reply to: [327871357](https://github.com/dotnet/roslyn/pull/38769#discussion_r327871357) [](ancestors = 327871357)

333fred replied:
No longer relevant.

---
In reply to: [327873967](https://github.com/dotnet/roslyn/pull/38769#discussion_r327873967) [](ancestors = 327873967,327871357)

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -129,16 +129,15 @@ public VisitArgumentResult(VisitResult visitResult, Optional<LocalState> stateFo
+        /// Records the updated Symbols from reparenting if the containing symbol was reinferred.
         /// </summary>
-        private readonly MethodSymbol _methodSignatureOpt;
+        private ReinferredVariableTracker _reinferredVariablesOpt;
```
>private ReinferredVariableTracker _reinferredVariablesOpt; [](start = 8, length = 58)

readonly?

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -383,6 +383,7 @@ protected override void Free()
                     _variableTypes.Add(key, value);
                 }
                 this.State = initialState.VariableNullableStates.Clone();
+                _reinferredVariablesOpt = reinferredVariablesOpt;
```
>_reinferredVariablesOpt = reinferredVariablesOpt [](start = 16, length = 48)

Already set above.

AlekseyTs replied:
> _reinferredVariablesOpt = reinferredVariablesOpt; [](start = 15, length = 50)

It looks like we already did this at the top.

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -1636,16 +1671,38 @@ public override BoundNode VisitLocalDeclaration(BoundLocalDeclaration node)
+            {
+                var updatedContainingSymbol = GetUpdatedContainingSymbol(node.LocalSymbol);
+                if (Symbol.Equals(node.LocalSymbol.ContainingSymbol, updatedContainingSymbol, TypeCompareKind.ConsiderEverything) &&
+                    (updatedType is null || node.LocalSymbol.TypeWithAnnotations.Equals(updatedType.Value, TypeCompareKind.ConsiderEverything)))
```
>node.LocalSymbol.TypeWithAnnotations.Equals(updatedType.Value, TypeCompareKind.ConsiderEverything) [](start = 44, length = 98)

Should we call `SetLocalType()` unconditionally, in case we're updating the type a second time, back to the original value?

333fred replied:
No longer relevant.

---
In reply to: [327874320](https://github.com/dotnet/roslyn/pull/38769#discussion_r327874320) [](ancestors = 327874320)

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -570,25 +570,26 @@ protected override ImmutableArray<PendingBranch> Scan(ref bool badRegion)
             Analyze(
                 compilation,
-                lambda.Symbol,
+                lambdaSymbolOpt ?? lambda.Symbol,
```
>lambdaSymbolOpt ?? lambda.Symbol [](start = 16, length = 32)

Are there cases when these two a different? If not, consider removing the ```lambdaSymbolOpt``` parameter.

333fred replied:
Currently, yes, we pass the updated LambdaSymbol through the `VisitConversion` chain.

---
In reply to: [327875276](https://github.com/dotnet/roslyn/pull/38769#discussion_r327875276) [](ancestors = 327875276)

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -668,12 +670,29 @@ private void TakeIncrementalSnapshot(BoundNode node)
+                updatedLocal = ((SourceLocalSymbol)originalLocal).WithContainingAndAnalyzedTypes(GetUpdatedContainingSymbol(originalLocal), updatedType);
+                if ((object)originalLocal == updatedLocal) return;
+                SetUpdatedSymbol(sourceNode, originalLocal, updatedLocal);
+                _reinferredVariablesOpt?.SetReinferredSymbol(originalLocal, updatedLocal);
```
>_reinferredVariablesOpt? [](start = 16, length = 24)

We already checked it for null

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -1509,6 +1539,10 @@ public override BoundNode VisitLocal(BoundLocal node)
             SetResult(node, GetAdjustedResult(type, slot), type);
+            if (_reinferredVariablesOpt?.GetReinferredSymbol(node.LocalSymbol) is LocalSymbol updated)
+            {
+                SetUpdatedSymbol(node, node.LocalSymbol, updated);
```
>node [](start = 33, length = 4)

The mapping does not depend on the node, we shouldn't be caching per node. And shouldn't be doing this on every VisitLocal.

333fred replied:
As discussed offline, it does for method calls, properties, and anything else where generics can be reinferred.

---
In reply to: [327882499](https://github.com/dotnet/roslyn/pull/38769#discussion_r327882499) [](ancestors = 327882499)

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -1636,16 +1671,38 @@ public override BoundNode VisitLocalDeclaration(BoundLocalDeclaration node)
+                    updatedType = node.LocalSymbol.TypeWithAnnotations;
+                }
+
+                SetLocalType(node, node.LocalSymbol, updatedType ?? node.LocalSymbol.TypeWithAnnotations);
```
>SetLocalType(node, node.LocalSymbol, updatedType ?? node.LocalSymbol.TypeWithAnnotations); [](start = 16, length = 90)

It looks like we are going to create symbols for this node for every pass we do. This feels like an overkill. I am probably missing something, but couldn't we just create symbols at the very end?

333fred replied:
Done.

---
In reply to: [327884012](https://github.com/dotnet/roslyn/pull/38769#discussion_r327884012) [](ancestors = 327884012)

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -3748,7 +3805,7 @@ private void CheckDisallowedNullAssignment(TypeWithState state, FlowAnalysisAnno
                         {
                             var varType = worstCaseParameterWithState.ToTypeWithAnnotations();
-                            _variableTypes[local.LocalSymbol] = varType;
+                            SetLocalType(local, local.LocalSymbol, varType);
```
>SetLocalType(local, local.LocalSymbol, varType); [](start = 28, length = 48)

Same comment as for VisitLocalDeclaration

333fred replied:
No longer applies as this code isn't changed.

---
In reply to: [327918526](https://github.com/dotnet/roslyn/pull/38769#discussion_r327918526) [](ancestors = 327918526)

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -5015,7 +5072,15 @@ private static BoundConversion GetConversionIfApplicable(BoundExpression convers
+                        if (delegateType is object)
+                        {
+                            var containingSymbol = GetUpdatedContainingSymbol(lambda.Symbol);
+                            updatedLambdaSymbol = lambda.CreateLambdaSymbol(delegateType, containingSymbol);
```
>updatedLambdaSymbol = lambda.CreateLambdaSymbol(delegateType, containingSymbol); [](start = 28, length = 80)

Same comment as for VisitLocalDeclaration

333fred replied:
No longer applies as this code isn't changed.

---
In reply to: [327918732](https://github.com/dotnet/roslyn/pull/38769#discussion_r327918732) [](ancestors = 327918732)

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -5553,8 +5618,11 @@ public override BoundNode VisitDelegateCreationExpression(BoundDelegateCreationE
                 case BoundLambda lambda:
                     {
-                        VisitLambda(lambda, delegateType, Diagnostics);
+                        var lambdaSymbol = lambda.CreateLambdaSymbol(delegateType, GetUpdatedContainingSymbol(lambda.Symbol));
```
>var lambdaSymbol = lambda.CreateLambdaSymbol(delegateType, GetUpdatedContainingSymbol(lambda.Symbol)); [](start = 24, length = 102)

Same comments as for VisitConversion

333fred replied:
No longer applies as this code isn't changed.

---
In reply to: [327919573](https://github.com/dotnet/roslyn/pull/38769#discussion_r327919573) [](ancestors = 327919573)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/LocalFunctionSymbol.cs:
```diff
@@ -503,9 +503,28 @@ public sealed override bool Equals(Symbol symbol, TypeCompareKind compareKind)
+            }
+            else
+            {
+                return new LocalFunctionSymbol(_binder, containingSymbol, _syntax);
```
> return new LocalFunctionSymbol(_binder, containingSymbol, _syntax); [](start = 15, length = 68)

It looks like this is going to force us to do binding of syntax in order to get all aspects of the local function. Attributes, parameter types, constraints, etc. I think we would want to avoid all that work. Perhaps by creating specializedwrapper symbols instead.

333fred replied:
I'm not entirely sure that's a bad thing here. Otherwise, we have to create a new system to recreate the local function symbol with all parameters substituted for reparented symbols, and the lazy init mechanism that will require.

---
In reply to: [327920751](https://github.com/dotnet/roslyn/pull/38769#discussion_r327920751) [](ancestors = 327920751)

333fred replied:
As discussed offline, I'm going to revert the LocalFunction changes in this PR entirely, to be addressed in a followup after equality is fixed.

---
In reply to: [329758111](https://github.com/dotnet/roslyn/pull/38769#discussion_r329758111) [](ancestors = 329758111,327920751)

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -5737,19 +5807,24 @@ public override BoundNode VisitLocalFunctionStatement(BoundLocalFunctionStatemen
                     snapshotBuilder = null;
                 }
 
+                LocalFunctionSymbol updatedLocalFunction = node.Symbol.WithContainingSymbol(GetUpdatedContainingSymbol(node.Symbol));
```
>LocalFunctionSymbol updatedLocalFunction = node.Symbol.WithContainingSymbol(GetUpdatedContainingSymbol(node.Symbol)); [](start = 16, length = 117)

Same comment about creating symbols for every pass.

333fred replied:
No longer being recreated every pass.

---
In reply to: [327923291](https://github.com/dotnet/roslyn/pull/38769#discussion_r327923291) [](ancestors = 327923291)

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -5774,6 +5849,10 @@ public override BoundNode VisitParameter(BoundParameter node)
             int slot = GetOrCreateSlot(parameter);
             var type = GetDeclaredParameterResult(parameter);
             SetResult(node, GetAdjustedResult(type, slot), type);
+            if (_reinferredVariablesOpt?.GetReinferredSymbol(node.ParameterSymbol) is ParameterSymbol updatedParameter)
```
>(_reinferredVariablesOpt?.GetReinferredSymbol(node.ParameterSymbol) is ParameterSymbol updatedParameter [](start = 15, length = 103)

I don't believe I saw where we add this information to the cache.

333fred replied:
No longer relevant.

---
In reply to: [327924175](https://github.com/dotnet/roslyn/pull/38769#discussion_r327924175) [](ancestors = 327924175)

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -6828,11 +6907,13 @@ public override void VisitForEachIterationVariables(BoundForEachStatement node)
-                            _variableTypes[iterationVariable] = result.ToTypeWithAnnotations();
+                            variableType = result.ToTypeWithAnnotations();
                         }
+                        SetLocalType(node, iterationVariable, variableType);
```
>SetLocalType(node, iterationVariable, variableType); [](start = 24, length = 52)

Is there a guarantee that we are dealing with "var" variable?

333fred replied:
No longer relevant.

---
In reply to: [327924666](https://github.com/dotnet/roslyn/pull/38769#discussion_r327924666) [](ancestors = 327924666)

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -1636,16 +1671,38 @@ public override BoundNode VisitLocalDeclaration(BoundLocalDeclaration node)
+                    updatedType = node.LocalSymbol.TypeWithAnnotations;
+                }
+
+                SetLocalType(node, node.LocalSymbol, updatedType ?? node.LocalSymbol.TypeWithAnnotations);
```
>SetLocalType(node, node.LocalSymbol, updatedType ?? node.LocalSymbol.TypeWithAnnotations); [](start = 16, length = 90)

Is there a guarantee that we are dealing with "var" variable?

333fred replied:
No longer relevant.

---
In reply to: [327924974](https://github.com/dotnet/roslyn/pull/38769#discussion_r327924974) [](ancestors = 327924974)

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:
```diff
@@ -699,9 +679,22 @@ internal override LocalSymbol GetAdjustedLocalSymbol(LocalSymbol local, int posi
         }
 
-        private LocalFunctionSymbol GetDeclaredLocalFunction(LocalFunctionStatementSyntax declarationSyntax, SyntaxToken declaredIdentifier)
+        private LocalFunctionSymbol GetDeclaredLocalFunction(LocalFunctionStatementSyntax declarationSyntax)
```
>GetDeclaredLocalFunction [](start = 36, length = 24)

The approach to adjusting symbols looks inconsistent. For a local we find the original symbol, then remap it. For a local function we are doing something different. I think we should have consistent approach for all local symbols (locals, local functions, parameters) and I am leaning towards approach used for locals.

AlekseyTs replied:
> The approach to adjusting symbols looks inconsistent. For a local we find the original symbol, then remap it. For a local function we are doing something different. I think we should have consistent approach for all local symbols (locals, local functions, parameters) and I am leaning towards approach used for locals.

This was marked as resolved, but I don't see any changes made.

---
In reply to: [327926167](https://github.com/dotnet/roslyn/pull/38769#discussion_r327926167) [](ancestors = 327926167)

AlekseyTs commented:
>         internal override LocalSymbol GetAdjustedLocalSymbol(LocalSymbol local, int position)

I don't believe I saw similar code for parameters.

---
Refers to: src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:666 in cc792be. [](commit_id = cc792bea0cfb536b626ab4541a69b4c5bfe14088, deletion_comment = False)

AlekseyTs commented:
>         internal override LocalSymbol GetAdjustedLocalSymbol(LocalSymbol local, int position)

What symbols do we get through Lookup APIs? We should be getting adjusted ones.

---
Refers to: src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:666 in cc792be. [](commit_id = cc792bea0cfb536b626ab4541a69b4c5bfe14088, deletion_comment = False)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs:
```diff
@@ -401,22 +401,24 @@ internal void SetTypeWithAnnotations(TypeWithAnnotations newType)
+        protected virtual SourceLocalSymbol CloneWithoutType(Symbol containingSymbol)
         {
-            return new SourceLocalSymbol(_containingSymbol, _scopeBinder, allowRefKind: false, _typeSyntax, _identifierToken, _declarationKind);
+            return new SourceLocalSymbol(containingSymbol, _scopeBinder, allowRefKind: false, _typeSyntax, _identifierToken, _declarationKind);
```
>return new SourceLocalSymbol(containingSymbol, _scopeBinder, allowRefKind: false, _typeSyntax, _identifierToken, _declarationKind); [](start = 12, length = 131)

I do not like the idea of cloning symbols that take binders, that means they potentially could binding something, but that should not happen. I think we should create thin wrappers that delegate to the original ones everything but container and type.

AlekseyTs commented:
Done with review pass (iteration 9).

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -5737,19 +5807,24 @@ public override BoundNode VisitLocalFunctionStatement(BoundLocalFunctionStatemen
                         snapshotBuilder,
                         returnTypesOpt: null);
+
+                SetUpdatedSymbol(node, node.Symbol, updatedLocalFunction);
```
>SetUpdatedSymbol(node, node.Symbol, updatedLocalFunction); [](start = 16, length = 58)

It feels like we would want to re-map references to type parameters of remapped local functions.

333fred replied:
[https://github.com/dotnet/roslyn/issues/38922](https://github.com/dotnet/roslyn/issues/38922) to track.

---
In reply to: [328236924](https://github.com/dotnet/roslyn/pull/38769#discussion_r328236924) [](ancestors = 328236924)

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:
```diff
@@ -699,9 +679,22 @@ internal override LocalSymbol GetAdjustedLocalSymbol(LocalSymbol local, int posi
         }
 
-        private LocalFunctionSymbol GetDeclaredLocalFunction(LocalFunctionStatementSyntax declarationSyntax, SyntaxToken declaredIdentifier)
+        private LocalFunctionSymbol GetDeclaredLocalFunction(LocalFunctionStatementSyntax declarationSyntax)
```
It feels like we would want to remap type parameters of local functions as well.

jcouv commented:
Marked PR as personal for now since Fred said he's reworking it.

333fred commented:
This is ready to be looked at again.

cston commented on src/Compilers/CSharp/Portable/Symbols/Wrapped/WrappedLocalSymbol.cs:
```diff
@@ -0,0 +1,43 @@
+    /// </summary>
+    internal abstract class WrappedLocalSymbol : LocalSymbol
+    {
+        public abstract LocalSymbol UnderlyingLocal { get; }
```
>public [](start = 8, length = 6)

`protected`?

cston commented on src/Compilers/CSharp/Portable/Symbols/UpdatedContainingSymbolLocal.cs:
```diff
@@ -0,0 +1,23 @@
+
+namespace Microsoft.CodeAnalysis.CSharp.Symbols
+{
+    internal sealed class UpdatedContainingSymbolLocal : WrappedLocalSymbol
```
>WrappedLocalSymbol [](start = 57, length = 18)

Is this the only implementation of `WrappedLocalSymbol`? If so, consider merging the two classes.

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.SnapshotManager.cs:
```diff
@@ -113,11 +133,42 @@ internal void VerifyNode(BoundNode node)
+                /// <remarks>
+                /// Lambda symbols are  mapped to the NameTypeSymbol of the delegate type they were reinferred to,
+                /// and are stored with a null node. The LambdaSymbol itself is position-independent, and does not
+                /// need any more information to server as a key.
```
>server [](start = 49, length = 6)

"serve"?

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.SnapshotManager.cs:
```diff
@@ -180,6 +232,30 @@ internal void TakeIncrementalSnapshot(BoundNode node, LocalState currentState)
+
+                internal void SetUpdatedSymbol(BoundNode node, Symbol originalSymbol, Symbol updatedSymbol)
+                {
+                    if (node is BoundLambda && originalSymbol is LambdaSymbol)
```
>node is BoundLambda && originalSymbol is LambdaSymbol [](start = 24, length = 53)

Please extract this to a helper method for use here and below.
```
_updatedSymbolMap[GetKey(node, originalSymbol)] =
    updatedSymbol;
```

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:
```diff
@@ -1656,22 +1656,51 @@ private void AppendSymbolsWithName(ArrayBuilder<Symbol> results, string name, Bi
+        {
+            if (!Compilation.NullableSemanticAnalysisEnabled) return symbol;
+
+            switch (symbol)
```
>switch (symbol) [](start = 12, length = 15)

It feels like we should also specially handle type parameters.

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:
```diff
@@ -1656,22 +1656,51 @@ private void AppendSymbolsWithName(ArrayBuilder<Symbol> results, string name, Bi
+                case ParameterSymbol _:
+                case MethodSymbol { MethodKind: MethodKind.LambdaMethod }:
+                case MethodSymbol { MethodKind: MethodKind.LocalFunction }:
+                    return RemapSymbolIfNecessaryCore(symbol, symbol.Locations[0].SourceSpan.Start);
```
>Locations[0] [](start = 69, length = 12)

Consider making code more robust by checking that we actually have a location. Perhaps asserting and skipping symbols that do not have locations.

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:
```diff
@@ -1656,22 +1656,51 @@ private void AppendSymbolsWithName(ArrayBuilder<Symbol> results, string name, Bi
+        /// Remaps a local, parameter, localfunction, or lambda symbol, if that symbol or its containing
+        /// symbols were reinferred. This should only be called when nullable semantic analysis is enabled.
+        /// </summary>
+        internal abstract Symbol RemapSymbolIfNecessaryCore(Symbol symbol, int position);
```
>int position [](start = 75, length = 12)

Consider removing this parameter and getting location where it is actually needed so that consumer wouldn't have to understand what position is appropriate to pass here.

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:
```diff
@@ -2233,6 +2242,30 @@ protected CSharpSyntaxNode GetBindableParentNode(CSharpSyntaxNode node)
             return bindableParent;
         }
 
+        internal override Symbol RemapSymbolIfNecessaryCore(Symbol symbol, int position)
```
>, int position [](start = 73, length = 14)

The parameter is not used, please remove it.

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:
```diff
@@ -2233,6 +2242,30 @@ protected CSharpSyntaxNode GetBindableParentNode(CSharpSyntaxNode node)
 
+        internal override Symbol RemapSymbolIfNecessaryCore(Symbol symbol, int position)
+        {
+            Debug.Assert(symbol is LocalSymbol ||
```
>symbol is LocalSymbol [](start = 25, length = 21)

I think we should also handle type parameter symbols.

333fred replied:
Tracked by [https://github.com/dotnet/roslyn/issues/38922](https://github.com/dotnet/roslyn/issues/38922)

---
In reply to: [329699547](https://github.com/dotnet/roslyn/pull/38769#discussion_r329699547) [](ancestors = 329699547)

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:
```diff
@@ -674,21 +670,14 @@ internal override LocalSymbol GetAdjustedLocalSymbol(LocalSymbol local, int posi
             {
-                if (!_analyzedVariableTypesOpt.TryGetValue(local, out adjustedLocal))
+                EnsureNullabilityAnalysisPerformedIfNecessary();
+                if (_remappedSymbols.TryGetValue(local, out var remappedLocal))
```
>_remappedSymbols [](start = 20, length = 16)

Below, in ```RemapSymbolIfNecessaryCore``` we null check this dictionary. Do we need to do the same here?

AlekseyTs commented on src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs:
```diff
@@ -1674,15 +1674,16 @@ internal static BoundBlock BindMethodBody(MethodSymbol method, TypeCompilationSt
                         // once.
                         // https://github.com/dotnet/roslyn/issues/35041
-                        methodBodyForSemanticModel = NullableWalker.AnalyzeAndRewrite(bodyBinder.Compilation, method, methodBody, bodyBinder, new DiagnosticBag(), createSnapshots: true, out snapshotManager);
+                        methodBodyForSemanticModel = NullableWalker.AnalyzeAndRewrite(bodyBinder.Compilation, method, methodBody, bodyBinder, new DiagnosticBag(), createSnapshots: true, out snapshotManager, out remappedSymbols);
```
>methodBodyForSemanticModel = NullableWalker.AnalyzeAndRewrite(bodyBinder.Compilation, method, methodBody, bodyBinder, new DiagnosticBag(), createSnapshots: true, out snapshotManager, out remappedSymbols); [](start = 24, length = 204)

Do we have an issue tracking making this work lazy?

333fred replied:
[https://github.com/dotnet/roslyn/issues/35041](https://github.com/dotnet/roslyn/issues/35041)

---
In reply to: [329704812](https://github.com/dotnet/roslyn/pull/38769#discussion_r329704812) [](ancestors = 329704812)

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.ReinferredVariableTracker.cs:
```diff
@@ -0,0 +1,67 @@
+    internal sealed partial class NullableWalker
+    {
+#nullable enable
+        internal sealed class ReinferredVariableTracker
```
>ReinferredVariableTracker [](start = 30, length = 25)

Is this class still used? Remove the file?

333fred replied:
It is not used, apparently when I added I changes I missed adding the delete.

---
In reply to: [329706888](https://github.com/dotnet/roslyn/pull/38769#discussion_r329706888) [](ancestors = 329706888)

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.SnapshotManager.cs:
```diff
@@ -180,6 +232,30 @@ internal void TakeIncrementalSnapshot(BoundNode node, LocalState currentState)
+                {
+                    if (node is BoundLambda && originalSymbol is LambdaSymbol)
+                    {
+                        _updatedSymbolMap[(null, originalSymbol)] = updatedSymbol;
```
>_updatedSymbolMap[(null, originalSymbol)] = updatedSymbol [](start = 24, length = 57)

Consider asserting that ```updatedSymbol``` is a named type of error type symbol.

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.SnapshotManager.cs:
```diff
@@ -180,6 +232,30 @@ internal void TakeIncrementalSnapshot(BoundNode node, LocalState currentState)
+                    }
+                    else
+                    {
+                        _updatedSymbolMap[(node, originalSymbol)] = updatedSymbol;
```
>_updatedSymbolMap[(node, originalSymbol)] = updatedSymbol [](start = 24, length = 57)

Consider asserting that ```originalSymbol``` and ```updatedSymbol``` are of the same Kind.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Wrapped/WrappedLocalSymbol.cs:
```diff
@@ -0,0 +1,43 @@
+        public override ImmutableArray<SyntaxReference> DeclaringSyntaxReferences => UnderlyingLocal.DeclaringSyntaxReferences;
+        public override string Name => UnderlyingLocal.Name;
+        public override bool IsImplicitlyDeclared => UnderlyingLocal.IsImplicitlyDeclared;
+        protected override Symbol OriginalSymbolDefinition => (object)UnderlyingLocal.OriginalDefinition == UnderlyingLocal ? this : UnderlyingLocal.OriginalDefinition;
```
>(object)UnderlyingLocal.OriginalDefinition == UnderlyingLocal ? this : UnderlyingLocal.OriginalDefinition; [](start = 62, length = 106)

I think this should always be ```this```.

333fred replied:
That's the default behavior, will remove this override.

---
In reply to: [329713644](https://github.com/dotnet/roslyn/pull/38769#discussion_r329713644) [](ancestors = 329713644)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Wrapped/WrappedLocalSymbol.cs:
```diff
@@ -0,0 +1,43 @@
+            UnderlyingLocal.GetConstantValueDiagnostics(boundInitValue);
+        internal override SyntaxNode GetDeclaratorSyntax() =>
+            UnderlyingLocal.GetDeclaratorSyntax();
+        internal override LocalSymbol WithSynthesizedLocalKindAndSyntax(SynthesizedLocalKind kind, SyntaxNode syntax) =>
```
>override [](start = 17, length = 8)

It feels like this API should be abstract because we would want to return an instance of a ```WrappedLocalSymbol```

333fred replied:
I actually debated making this throw completely. We should never be attempting to do this with a synthesized local.

---
In reply to: [329715544](https://github.com/dotnet/roslyn/pull/38769#discussion_r329715544) [](ancestors = 329715544)

333fred replied:
Proceeding with making this throw.

---
In reply to: [329828692](https://github.com/dotnet/roslyn/pull/38769#discussion_r329828692) [](ancestors = 329828692,329715544)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Wrapped/WrappedLocalSymbol.cs:
```diff
@@ -0,0 +1,43 @@
+    /// That behavior should be carefully reviewed and derived type
+    /// should override behavior as appropriate.
+    /// </summary>
+    internal abstract class WrappedLocalSymbol : LocalSymbol
```
>WrappedLocalSymbol [](start = 28, length = 18)

It feels like we would want to override Equals methods and GetHashCode in an abstract way.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs:
```diff
@@ -508,6 +486,7 @@ public sealed override bool Equals(Symbol obj, TypeCompareKind compareKind)
             var symbol = obj as SourceLocalSymbol;
             return (object)symbol != null
                 && symbol._identifierToken.Equals(_identifierToken)
+                && symbol.TypeWithAnnotations.Equals(TypeWithAnnotations, compareKind)
```
>&& symbol.TypeWithAnnotations.Equals(TypeWithAnnotations, compareKind) [](start = 16, length = 70)

When does this make a difference?

333fred replied:
It made a difference when we were updating SourceLocalSymbols. It no longer does.

---
In reply to: [329718695](https://github.com/dotnet/roslyn/pull/38769#discussion_r329718695) [](ancestors = 329718695)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs:
```diff
@@ -508,6 +486,7 @@ public sealed override bool Equals(Symbol obj, TypeCompareKind compareKind)
             var symbol = obj as SourceLocalSymbol;
```
>var symbol = obj as SourceLocalSymbol [](start = 12, length = 37)

Should we also handle UpdatedContainingSymbolLocal symbols here?

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/UpdatedContainingSymbolLocal.cs:
```diff
@@ -0,0 +1,23 @@
+{
+    internal sealed class UpdatedContainingSymbolLocal : WrappedLocalSymbol
+    {
+        internal UpdatedContainingSymbolLocal(LocalSymbol underlyingLocal, Symbol updatedContainingSymbol, TypeWithAnnotations updatedType)
```
>LocalSymbol [](start = 46, length = 11)

```SourceLocalSymbol```?

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -525,17 +502,25 @@ protected override ImmutableArray<PendingBranch> Scan(ref bool badRegion)
+
+        private static BoundNode Rewrite(ImmutableDictionary<BoundExpression, (NullabilityInfo, TypeSymbol)> updatedNullabilities, SnapshotManager snapshotManager, BoundNode node, out ImmutableDictionary<Symbol, Symbol> remappedSymbols)
+        {
+            var remappedSymbolsBuilder = ImmutableDictionary.CreateBuilder<Symbol, Symbol>(SymbolEqualityComparer.ConsiderEverything, SymbolEqualityComparer.ConsiderEverything);
```
>SymbolEqualityComparer.ConsiderEverything [](start = 91, length = 41)

Is this comparer based on ISymbol?

333fred replied:
Yes, but as far as I know we do not have an implementation of `IComparer<Symbol>`.

---
In reply to: [329722219](https://github.com/dotnet/roslyn/pull/38769#discussion_r329722219) [](ancestors = 329722219)

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -670,19 +646,23 @@ private void TakeIncrementalSnapshot(BoundNode node)
+                if (!Symbol.Equals(originalSymbol, updatedSymbol, TypeCompareKind.ConsiderEverything))
                 {
-                    _updatedSymbolMapOpt[key] = updatedSymbol;
+                    _snapshotBuilderOpt.SetUpdatedSymbol(node, originalSymbol, updatedSymbol);
```
>_snapshotBuilderOpt.SetUpdatedSymbol(node, originalSymbol, updatedSymbol); [](start = 20, length = 74)

It looks like we will always get here for a lambda and will never get into else for a lambda. Is this expected?

333fred replied:
No, will fix.

---
In reply to: [329723903](https://github.com/dotnet/roslyn/pull/38769#discussion_r329723903) [](ancestors = 329723903)

AlekseyTs commented:
>                 var parameterType = i >= signatureParameters.Length ? parameter.TypeWithAnnotations : signatureParameters[i].TypeWithAnnotations;

Is this code for using delegate parameter types for lambda parameters? I think it is not always appropriate to do that, lambda parameters can have explicit types.

---
Refers to: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:1337 in 46482d9. [](commit_id = 46482d9b636e0c8fd6eeb94d6cfd4f0f900cf470, deletion_comment = False)

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -1700,6 +1680,26 @@ bool canIgnoreAnonymousType(NamedTypeSymbol type)
+            static bool areLambdaAndNewDelegateSimilar(LambdaSymbol l, NamedTypeSymbol n)
+            {
+                var invokeMethod = n.DelegateInvokeMethod;
+                return invokeMethod.Parameters.SequenceEqual(l.Parameters,
```
>SequenceEqual [](start = 47, length = 13)

Count of parameters might not match in error scenarios.

333fred replied:
We don't have a test that fails this currently. Agreed that this can happen in error scenarios, but in that case we should probably drop the updated symbol, rather than record such a symbol, as that will observably change quickinfo and potentially IDE refactorings. I'll make a change to check just the lambda section when adding to the map, and abort if the lambda is different by parameter count.

---
In reply to: [329732324](https://github.com/dotnet/roslyn/pull/38769#discussion_r329732324) [](ancestors = 329732324)

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -5690,14 +5688,17 @@ private void VisitLambda(BoundLambda node, NamedTypeSymbol delegateTypeOpt, Diag
+            }
+
 
             Analyze(
```
>Analyze( [](start = 12, length = 8)

Extra empty line added above

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -6073,7 +6069,7 @@ private void VisitTupleDeconstructionArguments(ArrayBuilder<DeconstructionVariab
                     if (underlyingConversion.IsIdentity)
                     {
-                        if ((variable.Expression as BoundLocal)?.DeclarationKind == BoundLocalDeclarationKind.WithInferredType)
+                        if (((variable.Expression as BoundLocal)?.DeclarationKind == BoundLocalDeclarationKind.WithInferredType))
```
>if (((variable.Expression as BoundLocal)?.DeclarationKind == BoundLocalDeclarationKind.WithInferredType)) [](start = 24, length = 105)

Is this change necessary?

333fred replied:
No, the intent was to put it back as it was, I missed a paren.

---
In reply to: [329735618](https://github.com/dotnet/roslyn/pull/38769#discussion_r329735618) [](ancestors = 329735618)

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -7845,25 +7841,25 @@ public int GetHashCode((NullabilityInfo info, TypeSymbol type) obj)
             {
-                Debug.Assert(obj is ({ }, { }));
+                Debug.Assert(obj.sym is object);
                 return Hash.Combine(obj.expr, obj.sym.GetHashCode());
```
>obj.expr [](start = 36, length = 8)

Is it Ok to pass null here?

333fred replied:
Yes, it's appropriately nullchecked. Sam's PR annotates Hash.Combine here: [https://github.com/dotnet/roslyn/pull/38882/files#diff-a4f01f0c5d4e289dfae2bc84ec50366cR33](https://github.com/dotnet/roslyn/pull/38882/files#diff-a4f01f0c5d4e289dfae2bc84ec50366cR33)

---
In reply to: [329736470](https://github.com/dotnet/roslyn/pull/38769#discussion_r329736470) [](ancestors = 329736470)

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/NullabilityRewriter.cs:
```diff
@@ -74,18 +74,114 @@ private BoundNode VisitBinaryOperatorBase(BoundBinaryOperatorBase binaryOperator
             }
 
-            return sym;
+            return (T)updatedSymbol;
```
>(T)updatedSymbol [](start = 19, length = 16)

Should we rewrite references to type parameters of local functions?

333fred replied:
Yes, but this is tracked by [https://github.com/dotnet/roslyn/issues/38922](https://github.com/dotnet/roslyn/issues/38922).

---
In reply to: [329749865](https://github.com/dotnet/roslyn/pull/38769#discussion_r329749865) [](ancestors = 329749865)

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/NullabilityRewriter.cs:
```diff
@@ -74,18 +74,114 @@ private BoundNode VisitBinaryOperatorBase(BoundBinaryOperatorBase binaryOperator
+                if (updatedDelegateType is null)
+                {
+                    Debug.Assert(updatedContaining is object);
+                    var parameterRefKinds = lambda.ParameterRefKinds.IsDefault ? ImmutableArray<RefKind>.Empty : lambda.ParameterRefKinds;
```
>var parameterRefKinds = lambda.ParameterRefKinds.IsDefault ? ImmutableArray<RefKind>.Empty : lambda.ParameterRefKinds; [](start = 20, length = 118)

It feels strange that CreateLambdaSymbol doesn't just take ParameterRefKinds from the original lambda and requires them to be adjusted in some way. If that is really needed, It would probably make sense to encapsulate the adjustment inside the API.

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/NullabilityRewriter.cs:
```diff
@@ -74,18 +74,114 @@ private BoundNode VisitBinaryOperatorBase(BoundBinaryOperatorBase binaryOperator
+                    updatedLambda = boundLambda.CreateLambdaSymbol(updatedDelegateType, updatedContaining ?? lambda.ContainingSymbol);
+                }
+
+                _remappedSymbols.Add(lambda, updatedLambda);
```
> _remappedSymbols.Add(lambda, updatedLambda); [](start = 15, length = 45)

Is there a guarantee that we will not try to do this more than once?

333fred replied:
`Add` will throw if we add attempt to add the same key twice.

---
In reply to: [329752539](https://github.com/dotnet/roslyn/pull/38769#discussion_r329752539) [](ancestors = 329752539)

AlekseyTs replied:
> Add will throw if we add attempt to add the same key twice.

Is there a guarantee that this will not happen?

---
In reply to: [329827877](https://github.com/dotnet/roslyn/pull/38769#discussion_r329827877) [](ancestors = 329827877,329752539)

333fred replied:
Nothing beyond the standard invariants around how BoundLambdas generate their symbols when asked for them, and that such symbols are unique.

---
In reply to: [330680758](https://github.com/dotnet/roslyn/pull/38769#discussion_r330680758) [](ancestors = 330680758,329827877,329752539)

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/NullabilityRewriter.cs:
```diff
@@ -74,18 +74,114 @@ private BoundNode VisitBinaryOperatorBase(BoundBinaryOperatorBase binaryOperator
+                    _remappedSymbols.Add(localFunction.Parameters[i], updatedLocalFunction.Parameters[i]);
+                }
+
+                return updatedLocalFunction;
```
>return updatedLocalFunction; [](start = 16, length = 28)

I think we need to add remapping for type parameters as well.

333fred replied:
Tracked by [https://github.com/dotnet/roslyn/issues/38922](https://github.com/dotnet/roslyn/issues/38922).

---
In reply to: [329753419](https://github.com/dotnet/roslyn/pull/38769#discussion_r329753419) [](ancestors = 329753419)

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/NullabilityRewriter.cs:
```diff
@@ -96,15 +192,16 @@ private BoundNode VisitBinaryOperatorBase(BoundBinaryOperatorBase binaryOperator
             foreach (var originalSymbol in symbols)
             {
-                if (_updatedSymbols.TryGetValue((expr, originalSymbol), out var updatedSymbol))
+                if (originalSymbol is null) continue;
```
>if (originalSymbol is null) continue; [](start = 16, length = 37)

Are we going to drop the null from the result?

333fred replied:
Good catch.

---
In reply to: [329754375](https://github.com/dotnet/roslyn/pull/38769#discussion_r329754375) [](ancestors = 329754375)

AlekseyTs commented:
Done with review pass (iteration 11)

333fred commented:
>                 var parameterType = i >= signatureParameters.Length ? parameter.TypeWithAnnotations : signatureParameters[i].TypeWithAnnotations;

The only time `_delegateInvokeMethod` is non-null is when there aren't explicit type parameters.

---
In reply to: [536697231](https://github.com/dotnet/roslyn/pull/38769#issuecomment-536697231) [](ancestors = 536697231)

---
Refers to: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:1337 in 46482d9. [](commit_id = 46482d9b636e0c8fd6eeb94d6cfd4f0f900cf470, deletion_comment = False)

AlekseyTs commented:
@333fred It looks like the build is broken.

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs:
```diff
@@ -126,6 +125,23 @@ public TypeWithAnnotations GetInferredReturnType(ConversionsBase conversions, Nu
+                returnType,
+                diagnostics: null,
+                parameterTypes,
+                parameterRefKinds.IsDefault ? ImmutableArray.CreateRange(parameterTypes, _ => RefKind.None) : parameterRefKinds,
```
>parameterTypes, _ => RefKind.None [](start = 73, length = 33)

Consider passing Enumerable.Repeat instead to avoid lambdas altogether.

AlekseyTs commented:
>         internal override LocalSymbol GetAdjustedLocalSymbol(LocalSymbol local, int position)

> It looks like this parameter is no longer used. Remove?

This was marked as resolved, but the parameter is still present.

---
In reply to: [534686358](https://github.com/dotnet/roslyn/pull/38769#issuecomment-534686358) [](ancestors = 534686358)

---
Refers to: src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:666 in 708099d. [](commit_id = 708099da280d909e093be09cca26646f5b8412f4, deletion_comment = False)

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/SyntaxTreeSemanticModel.cs:
```diff
@@ -2234,5 +2234,24 @@ public override DeconstructionInfo GetDeconstructionInfo(ForEachVariableStatemen
+        {
+            var location = symbol switch
+            {
+                LocalSymbol { Locations: { IsDefaultOrEmpty: false } locations } => locations[0],
```
>{ Locations: { IsDefaultOrEmpty: false } locations } [](start = 28, length = 52)

This code is completed three times. This check could be performed once. It doesn't have to be bundles with symbol kind check.

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -1700,8 +1714,28 @@ bool canIgnoreAnonymousType(NamedTypeSymbol type)
+        }
 #endif
 
+        private static bool AreLambdaAndNewDelegateSimilar(LambdaSymbol l, NamedTypeSymbol n, bool exactMatch = false)
```
>, bool exactMatch = false [](start = 92, length = 25)

It looks like added parameter is not used.

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -1341,14 +1355,14 @@ protected override LocalState ReachableBottomState()
 
             var methodParameters = methodSymbol.Parameters;
-            var signatureParameters = _useMethodSignatureParameterTypes ? _methodSignatureOpt.Parameters : methodParameters;
+            var signatureParameters = (_useMethodSignatureParameterTypes ? _delegateInvokeMethod : methodSymbol).Parameters;
```
>_useMethodSignatureParameterTypes [](start = 39, length = 33)

Is there a guarantee that ```_delegateInvokeMethod``` is not null when ```_useMethodSignatureParameterTypes``` is true. Constructor used to ensure that before, but no longer does this, and I think there was another place in the code where ```_delegateInvokeMethod``` was explicitly checked for null in addition to ```_useMethodSignatureParameterTypes```.

333fred replied:
Added assert back.

---
In reply to: [330704941](https://github.com/dotnet/roslyn/pull/38769#discussion_r330704941) [](ancestors = 330704941)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/UpdatedContainingSymbolLocal.cs:
```diff
@@ -0,0 +1,69 @@
+
+        public override int GetHashCode()
+        {
+            return Hash.Combine(UnderlyingLocal,
```
>return Hash.Combine(UnderlyingLocal, [](start = 12, length = 36)

I think we should simply delegate to the ```UnderlyingLocal```. Under some condition this symbol should be considered equal to the ```UnderlyingLocal```, therefore both should have the same hash code.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/UpdatedContainingSymbolLocal.cs:
```diff
@@ -0,0 +1,69 @@
+                return true;
+            }
+
+            return other is LocalSymbol local
```
>return other is LocalSymbol local [](start = 12, length = 33)

It is hard to tell if this is giving us correct behavior. I would separately handle a case when ```other``` is ```SourceLocalSymbol``` and when ```other``` is ```UpdatedContainingSymbolAndNullableAnnotationLocal```. For all other types the result should be false. Then, I would compare two ```SourceLocalSymbol``` for the exact match, and only then would compare type and container with the given ```compareKind```.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/LocalFunctionSymbol.cs:
```diff
@@ -503,9 +503,9 @@ public sealed override bool Equals(Symbol symbol, TypeCompareKind compareKind)
-            var localFunction = symbol as LocalFunctionSymbol;
-            return (object)localFunction != null
-                && localFunction._syntax == _syntax;
+            return symbol is LocalFunctionSymbol localFunction
```
>return symbol is LocalFunctionSymbol localFunction [](start = 12, length = 50)

It is not clear why this code is changed. Consider reverting the file to its original state.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/UpdatedContainingSymbolLocal.cs:
```diff
@@ -0,0 +1,69 @@
+                && local.Equals(UnderlyingLocal, TypeCompareKind.AllNullableIgnoreOptions | TypeCompareKind.IgnoreTupleNames);
+        }
+
+        public override int GetHashCode()
```
>GetHashCode [](start = 28, length = 11)

Please make sure to have targeted tests for the GetHashCode\Equals covering matrix: 
- SourceLocalSymbol to UpdatedContainingSymbolAndNullableAnnotationLocal
- UpdatedContainingSymbolAndNullableAnnotationLocal to SourceLocalSymbol 
- UpdatedContainingSymbolAndNullableAnnotationLocal to UpdatedContainingSymbolAndNullableAnnotationLocal

AlekseyTs commented:
Done with review pass (iteration 13)

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [537638309](https://github.com/dotnet/roslyn/pull/38769#issuecomment-537638309) [](ancestors = 537638309)

333fred commented:
Note: compilation failures are because of issues that I will need to adjust to after merging from master, with the changes introduced in https://github.com/dotnet/roslyn/pull/38750. I thought that those would be straight up merge conflicts, but it turns out that git was able to merge successfully, it just doesn't build.

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/Symbols/Source/UpdatedContainingSymbolAndNullableAnntotationTests.cs:
```diff
@@ -0,0 +1,68 @@
+            assertEquality(wrappedLocal1, wrappedLocal1, nullableIgnored: true, considerEverything: true);
+            assertEquality(wrappedLocal1, wrappedLocal1a, nullableIgnored: true, considerEverything: true);
+
+            void assertEquality(Symbol symbol1, Symbol symbol2, bool nullableIgnored, bool considerEverything)
```
>nullableIgnored, bool considerEverything [](start = 69, length = 40)

Consider asserting that, if ```considerEverything``` is true, then ```nullableIgnored``` is true. Or explicitly make an adjustment to flags at the beginning of the function.

AlekseyTs commented:
Done with review pass (iteration 14). Would like to see CI pass before signing off.

333fred commented:
@cston @AlekseyTs please review.

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/NullabilityRewriter.cs:
```diff
@@ -76,16 +76,87 @@ private BoundNode VisitBinaryOperatorBase(BoundBinaryOperatorBase binaryOperator
-                return (T)updatedSymbol;
+                updatedSymbol = sym;
             }
+            // On merge, replace with RoslynDebug.Assert and remove suppression
```
>// On merge, replace with RoslynDebug.Assert and remove suppression [](start = 12, length = 67)

It looks like the comment is obsolete now.

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs:
```diff
@@ -638,33 +673,29 @@ private void ValidateUnsafeParameters(DiagnosticBag diagnostics, ImmutableArray<
+            return BindWithDelegateAndReturnType(cacheKey.ParameterTypes, cacheKey.ParameterRefKinds, returnType);
         }
 
         private (LambdaSymbol lambdaSymbol, BoundBlock block, ExecutableCodeBinder lambdaBodyBinder, DiagnosticBag diagnostics) BindWithDelegateAndReturnType(
```
>BindWithDelegateAndReturnType [](start = 128, length = 29)

The name no longer reflects the input for the function. Consider adjusting.

cston commented on src/Compilers/CSharp/Portable/BoundTree/NullabilityRewriter.cs:
```diff
@@ -76,16 +76,87 @@ private BoundNode VisitBinaryOperatorBase(BoundBinaryOperatorBase binaryOperator
+            {
+                var updatedDelegateType = _snapshotManager?.GetUpdatedDelegateTypeForLambda(lambda);
+
+                if (!_remappedSymbols.TryGetValue(lambda.ContainingSymbol, out Symbol? updatedContaining) && updatedDelegateType is null)
```
>updatedDelegateType is null [](start = 109, length = 27)

Consider checking this first since this is cheap.

cston commented on src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:
```diff
@@ -660,49 +667,45 @@ private LocalSymbol GetDeclaredLocal(CSharpSyntaxNode declarationSyntax, SyntaxT
+        private LocalFunctionSymbol GetDeclaredLocalFunction(LocalFunctionStatementSyntax declarationSyntax)
+        {
+            // If nullable semantic analysis is enabled, we need to potentially reinfer the containing symbol
+            // of a local function, as it could be in a reinferred lambda. For an example, see
```
The comment makes sense but it doesn't look like we're explicitly reinferring the containing symbol here. (That just happens to be a necessary part of reinferring the local function.) Consider removing the comment.

cston commented on src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:
```diff
@@ -660,49 +667,45 @@ private LocalSymbol GetDeclaredLocal(CSharpSyntaxNode declarationSyntax, SyntaxT
+            EnsureNullabilityAnalysisPerformedIfNecessary();
+            if (_lazyRemappedSymbols is null) return originalSymbol;
+
+            if (_lazyRemappedSymbols.TryGetValue(originalSymbol, out Symbol remappedSymbol))
```
Minor point: Consider merging to `if (_lazyRemappedSymbols?.TryGetValue(...) == true)`.

333fred replied:
I specifically didn't do this because definite assignment cannot figure out that `remappedSymbol` is definitely assigned by that test.

---
In reply to: [331618450](https://github.com/dotnet/roslyn/pull/38769#discussion_r331618450) [](ancestors = 331618450)

cston commented on src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:
```diff
@@ -2233,6 +2248,29 @@ protected CSharpSyntaxNode GetBindableParentNode(CSharpSyntaxNode node)
+                return symbol;
+            }
+
+            if (_lazyRemappedSymbols.TryGetValue(symbol, out var remappedSymbol))
```
`if (_lazyRemappedSymbols?.TryGetValue(...) == true)`

333fred replied:
Same comment about definite assignment.

---
In reply to: [331619521](https://github.com/dotnet/roslyn/pull/38769#discussion_r331619521) [](ancestors = 331619521)

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -670,19 +669,34 @@ private void TakeIncrementalSnapshot(BoundNode node)
         private void SetUpdatedSymbol(BoundNode node, Symbol originalSymbol, Symbol updatedSymbol)
         {
-            if (_updatedSymbolMapOpt is object)
+            var lambdaisExactMatch = false;
```
>lambdaisExactMatch [](start = 16, length = 18)

`lambdaIsExactMatch` (uppercase `i`)

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -670,19 +669,34 @@ private void TakeIncrementalSnapshot(BoundNode node)
+            Debug.Assert(AreCloseEnough(originalSymbol, updatedSymbol), $"Attempting to set {node.Syntax} from {originalSymbol.ToDisplayString()} to {updatedSymbol.ToDisplayString()}");
+#endif
+
+            if (_snapshotBuilderOpt is object)
```
>if (_snapshotBuilderOpt is object) [](start = 12, length = 34)

Return early (before checks above) if `_snapshotBuilderOpt is null`.

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -670,19 +669,34 @@ private void TakeIncrementalSnapshot(BoundNode node)
+
+            if (_snapshotBuilderOpt is object)
+            {
+                if (Symbol.Equals(originalSymbol, updatedSymbol, TypeCompareKind.ConsiderEverything) || lambdaisExactMatch)
```
>lambdaisExactMatch [](start = 104, length = 18)

Consider checking this first since it's cheap.

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.SnapshotManager.cs:
```diff
@@ -113,11 +134,42 @@ internal void VerifyNode(BoundNode node)
+                /// rewriter phase of the compiler.
+                /// </summary>
+                /// <remarks>
+                /// Lambda symbols are  mapped to the NameTypeSymbol of the delegate type they were reinferred to,
```
>   [](start = 38, length = 2)

Extra space.

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.SnapshotManager.cs:
```diff
@@ -180,6 +233,31 @@ internal void TakeIncrementalSnapshot(BoundNode node, LocalState currentState)
+                    _updatedSymbolMap.Remove(GetKey(node, symbol));
+                }
+
+                private (BoundNode?, Symbol) GetKey(BoundNode node, Symbol symbol)
```
>private [](start = 16, length = 7)

`static`

cston commented on src/Compilers/CSharp/Portable/Symbols/UpdatedContainingSymbolLocal.cs:
```diff
@@ -0,0 +1,85 @@
+            TypeWithAnnotations = updatedType;
+        }
+
+        private SourceLocalSymbol UnderlyingLocal { get; }
```
>private SourceLocalSymbol UnderlyingLocal { get; } [](start = 8, length = 50)

`private readonly SourceLocalSymbol _underlyingLocal;`

cston commented on src/Compilers/CSharp/Portable/Symbols/UpdatedContainingSymbolLocal.cs:
```diff
@@ -0,0 +1,85 @@
+        internal UpdatedContainingSymbolAndNullableAnnotationLocal(SourceLocalSymbol underlyingLocal, Symbol updatedContainingSymbol, TypeWithAnnotations updatedType)
+        {
+            Debug.Assert(underlyingLocal is object);
+            Debug.Assert(updatedContainingSymbol is object);
```
>Debug.Assert(updatedContainingSymbol is object); [](start = 12, length = 48)

```
Debug.Assert(updatedContainingSymbol.Equals(
    underlyingLocal.ContainingSymbol,
    TypeCompareKind.AllNullableIgnoreOptions));
```

cston commented on src/Compilers/CSharp/Portable/Symbols/UpdatedContainingSymbolLocal.cs:
```diff
@@ -0,0 +1,85 @@
+            SourceLocalSymbol? otherSource = otherLocal switch
+            {
+                UpdatedContainingSymbolAndNullableAnnotationLocal updated => updated.UnderlyingLocal,
+                SourceLocalSymbol source => source,
```
>SourceLocalSymbol [](start = 16, length = 17)

Treating `UpdatedContainingSymbolLocal` and `SourceLocalSymbol` equal seems fragile.

What was the reason we chose not to clone `SourceLocalSymbol` with updated `ContainingSymbol` and `Type`?

333fred replied:
Cloning it would necessitate passing a binder around and potentially causing new binding to occur, and it's easier to reason about this smaller wrapper class instead of what will happen when we create a full new symbol.
Additionally, this class will likely serve as a starting point for a wrapped ITypeSymbol later when Aleksey splits the tree.

---
In reply to: [331689353](https://github.com/dotnet/roslyn/pull/38769#discussion_r331689353) [](ancestors = 331689353)

333fred commented:
@cston addressed feedback.
